aspnetcore/fundamentals/app-state.md

```diff
@@ -187,37 +201,34 @@ The following code displays `TempData["Message"]`, but at the end of the request
+* Does the app already use session state? If so, using the session state TempData provider has no additional cost to the app beyond the size of the data.
+* Does the app use TempData only sparingly for relatively small amounts of data, up to 500 bytes? If so, the cookie TempData provider adds a small cost to each request that carries TempData. If not, the session state TempData provider can be beneficial to avoid round-tripping a large amount of data in each request until the TempData is consumed.
+* Does the app run in a server farm on multiple servers? If so, there's no additional configuration required to use the cookie TempData provider outside of Data Protection (see <xref:security/data-protection/introduction> and [Key storage providers](xref:security/data-protection/implementation/key-storage-providers)).
 
-> [!NOTE]
-> Most web clients (such as web browsers) enforce limits on the maximum size of each cookie, the total number of cookies, or both. When using the cookie TempData provider, verify the app won't exceed these limits. Consider the total size of the data. Account for increases in cookie size due to encryption and chunking.
+Most web clients such as web browsers enforce limits on the maximum size of each cookie and the total number of cookies. When using the cookie TempData provider, verify the app won't exceed [these limits](http://www.faqs.org/rfcs/rfc2965.html). Consider the total size of the data. Account for increases in cookie size due to encryption and chunking.
 
 ### Configure the TempData provider
 
 The cookie-based TempData provider is enabled by default.
 
 To enable the session-based TempData provider, use the [AddSessionStateTempDataProvider](/dotnet/api/microsoft.extensions.dependencyinjection.mvcviewfeaturesmvcbuilderextensions.addsessionstatetempdataprovider) extension method:
 
-[!code-csharp[](app-state/samples_snapshot_2/2.x/SessionSample/Startup.cs?name=snippet1&highlight=11,13,32)]
+<!-- requires review: -->
+[!code-csharp[](app-state/samples/3.x/SessionSample/Startup3.cs?name=snippet1&highlight=4,6,25)]
 
-The order of middleware is important. In the preceding example, an `InvalidOperationException` exception occurs when `UseSession` is invoked after `UseMvc`. For more information, see [Middleware Ordering](xref:fundamentals/middleware/index#order).
-
-> [!IMPORTANT]
-> If targeting .NET Framework and using the session-based TempData provider, add the [Microsoft.AspNetCore.Session](https://www.nuget.org/packages/Microsoft.AspNetCore.Session/) package to the project.
+The order of middleware is important. Call `UseSession` after `UseStaticFiles` and before `UseRouting`. See [Middleware Ordering](xref:fundamentals/middleware/index#order).
```

Comment:
A couple of comments here:

- The referenced "Middleware order" section shows `UseSession` _after_ `UseRouting`, but before `UseEndpoints`. I think that's correct.
- This explanation appears twice in the topic. Just flagging that in case it's not intentional.